text/0098-store-multiple-replay-segments-in-a-single-blob.md

```diff
@@ -0,0 +1,282 @@
+
+1. We download the file blob of the compromised DEK (there will only be one).
+2. We decrypt the subset of bytes related to the compromised DEK.
+3. We generate a new DEK.
+4. We encrypt the decrypted bytes with the new DEK.
+5. We encrypt the new DEK with any KEK.
+6. We stitch the blob back together and upload to cloud storage.
+7. We re-write the metadata rows.
+   - New offsets are written because encryption is not guaranteed to produce the same length of bytes.
+   - The compromised DEK is overwritten during this step.
+
+**A Word on the Necessity of Key Rotation**
+
+What's the likelihood of a DEK or KEK leaking? Probably small but larger than you think.
+
+What happens if an exception is raised while trying to read a file? The sentry_sdk will sweep in the `locals()` from the application state. This will include KEKs, DEKs, filenames, and byte-ranges. Everything an attacker needs to read private data. This data will be presented in a human-readable format on the Sentry.io web app. This private data may even find its way into our logs (in whole or in part) providing another vector of attack.
+
+Now, the value of the data is very small. Its not "worth" stealing. After all, we're not encrypting this data now. We're only adopting encryption to "delete" files in clever way.
+
+However, the security and privacy of these keys is critical for us to maintain our "deletion" guarantees to customers and to regulators. Currently, when we say a file is deleted it is in fact deleted. Going forward it will still exist but will be impossible to read. That un-readability promise _has to be maintained_. Compromised keys must be rotated and strong patterns for doing so must be maintained.
+
+# Drawbacks
+
+- Data is only deleted after the retention period.
+  - This means customer data must be protected between the deletion date and retention expiry.
+
+# Questions
+
+1. If KEKs are managed in a remote service how do we manage outages?
+   - We manage it in the same way we manage an outage of any other remote service.
+     - We will need to backlog or otherwise 400/404.
+   - KEKs have the benefit of _probably_ not blocking ingest as the key will be cached for long stretches of time (24 hours) and can be used for longer periods of time if a new key can not be fetched.
+2. How will read efficiency be impacted if we rely on a remote service to decrypt blob data?
+   - It will have some cost but hopefully that cost is minimized by the constraints of your system.
+   - For example, Session Replay fetches multiple segment blobs in a single request. At most we will need to fetch two keys (and in the majority of cases a single key) to decrypt the segments.
+   - This key fetching latency is immaterial to the total latency of the request.
```

Comment:
What is the self hosted open source implementation for this

Reply:
Environment variable named `SECRET`.